contracts/InstallationDiamond/facets/InstallationUpgradeFacet.sol

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import {LibAppStorageInstallation, InstallationType, QueueItem, UpgradeQueue, Modifiers} from "../../libraries/AppStorageInstallation.sol";
+import {LibSignature} from "../../libraries/LibSignature.sol";
+import {RealmDiamond} from "../../interfaces/RealmDiamond.sol";
+import {IERC721} from "../../interfaces/IERC721.sol";
+import {IERC20} from "../../interfaces/IERC20.sol";
+import {LibItems} from "../../libraries/LibItems.sol";
+import {InstallationAdminFacet} from "./InstallationAdminFacet.sol";
+import {LibInstallation} from "../../libraries/LibInstallation.sol";
+import {LibERC1155} from "../../libraries/LibERC1155.sol";
+
+contract InstallationUpgradeFacet is Modifiers {
+  event UpgradeInitiated(
+    uint256 indexed _realmId,
+    uint256 _coordinateX,
+    uint256 _coordinateY,
+    uint256 blockInitiated,
+    uint256 readyBlock,
+    uint256 installationId
+  );
+
+  event UpgradeFinalized(uint256 indexed _realmId, uint256 _coordinateX, uint256 _coordinateY, uint256 _newInstallationId);
+
+  event UpgradeQueued(address indexed _owner, uint256 indexed _realmId, uint256 indexed _queueIndex);
+  event UpgradeQueueFinalized(address indexed _owner, uint256 indexed _realmId, uint256 indexed _queueIndex);
+
+  /// @notice Allow a user to upgrade an installation in a parcel
+  /// @dev Will throw if the caller is not the owner of the parcel in which the installation is installed
+  /// @param _upgradeQueue A struct containing details about the queue which contains the installation to upgrade
+  ///@param _signature API signature
+  ///@param _gltr Amount of GLTR to use, can be 0
+  function upgradeInstallation(
+    UpgradeQueue calldata _upgradeQueue,
+    bytes memory _signature,
+    uint40 _gltr
+  ) external {
+    require(
+      LibSignature.isValid(
+        keccak256(abi.encodePacked(_upgradeQueue.parcelId, _upgradeQueue.coordinateX, _upgradeQueue.coordinateY, _upgradeQueue.installationId)),
+        _signature,
+        s.backendPubKey
+      ),
+      "InstallationAdminFacet: Invalid signature"
+    );
+    // check owner
+    require(IERC721(s.realmDiamond).ownerOf(_upgradeQueue.parcelId) == _upgradeQueue.owner, "InstallationFacet: Not owner");
+    // check coordinates
+    RealmDiamond realm = RealmDiamond(s.realmDiamond);
+
+    //check upgradeQueueCapacity
+    require(
+      realm.getParcelUpgradeQueueCapacity(_upgradeQueue.parcelId) > realm.getParcelUpgradeQueueLength(_upgradeQueue.parcelId),
+      "InstallationFacet: UpgradeQueue full"
+    );
+
+    realm.checkCoordinates(_upgradeQueue.parcelId, _upgradeQueue.coordinateX, _upgradeQueue.coordinateY, _upgradeQueue.installationId);
+
+    // check unique hash
+    bytes32 uniqueHash = keccak256(
+      abi.encodePacked(_upgradeQueue.parcelId, _upgradeQueue.coordinateX, _upgradeQueue.coordinateY, _upgradeQueue.installationId)
```

Comment:
This should be changed to regular encode since this is checking for uniqueness of the upgrade

Reply:
We'll need to write some tests to ensure that changing this does not break any current upgrades